### DISCUSSION TITLE
Conversion of string to int in C/C++ without using stdlib module

### DISCUSSION THREAD

**Discussion by Sarvang-Yadava** [community/community#137018] [Programming Help] [answered]
### Body

I wanted to convert a string like "1" to integers but it's showing that you cannot convert a bigger size element to smaller.

And I know atoi can do this but I wanted to know how?

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by Giddy-K** [comment, marked as answer]
Converting a string to an integer without using the stdlib library can be done by manually processing each character of the string. Here's how you can do it:

    1. Initialize a Result Variable: Start with an integer variable result initialized to 0.
    2. Process Each Character: Iterate through each character of the string.
    3. Convert Each Character to an Integer: Subtract '0' from the character to get its integer value.
    4. Update the Result: Multiply the current result by 10 (to shift it left by one digit) and add the integer value of the current character.
    5. Handle Negative Numbers: If the string starts with a '-', store this information and convert the rest of the string as usual. Then, multiply the final result by -1.

Example Code in C/C++:

```
#include <iostream>
#include <cstring>  // Only for strlen (optional)

int stringToInt(const char* str) {
    int result = 0;
    int sign = 1;
    int i = 0;

    // Handle negative numbers
    if (str[0] == '-') {
        sign = -1;
        i++;  // Skip the '-' character
    }

    // Iterate over each character
    for (; str[i] != '\0'; ++i) {
        // Convert character to integer and add to result
        result = result * 10 + (str[i] - '0');
    }

    return sign * result;
}

int main() {
    const char* str = "-1234";
    int num = stringToInt(str);
    std::cout << "Converted number: " << num << std::endl;

    return 0;
}
```

Explanation:

-    sign = -1: If the string starts with a '-', set the sign to -1 and skip the first character.

-     result = result * 10 + (str[i] - '0'): This line converts the character to its integer equivalent and accumulates it in the result. For example, if the current character is '3', str[i] - '0' will give 3.

-     Returning sign * result: If the string was negative, multiply the final result by -1.

Handling Edge Cases:

-     Empty String: If the input string is empty, the function might need additional checks.

-     Invalid Characters: If the string contains non-numeric characters, you'd typically add error handling.

-     Overflow/Underflow: For large numbers, the result may overflow. Handling this would require additional checks.

This method provides a manual way to convert a string to an integer, simulating what functions like atoi do internally.

**Reply by Sarvang-Yadava** [reply]
So str[i] is seen as a character already and by subtracting '0', we just converting every single char to integer.
Then we multiplying 10 to result everytime to increase the places like units place tens place etc... right?

**Reply by Giddy-K** [reply]
Exactly...

**Reply by Sarvang-Yadava** [reply]
Thankss

**Reply by Giddy-K** [reply]
You're welcome
